Remove type expectation from IN

Mats-SX opened the pull request:
Fixes #5921

Mats-SX committed:
Make IN always expect Collection<Any>

Previously `IN` tried to be clever and calculate its expected right-hand
side type by deriving from the left-hand side. This does not work when
the right-hand side has an integer of type CTNumber. From now on, `IN`
will always expect its right-hand side to be any collection.

Fixes #5921

cleishm commented:
Removing the type expectation seems heavy handed, especially as the expectation works ok with the rule planner or when parameters are substituted in: which means the expectation is ok. I assume it was also there in 2.2 which was ok?

craigtaverner commented:
Chris, the issue started in 2.2.3. In 2.2.2 and before there was no expectation, mainly because a third round of semantic checking was only added in 2.2.3, and that round occurs after rewriting id(n)=x to id(n) in [x] and it turned out the type checking for IN is much more stringent than for '='. The PR is a proposal to be discussed (hence the 'not ready for merge'). We would prefer that type checking allow Int to be compatible with [Number], but it appears not to be the case (in either direction). Neither casting nor coercion is supported, and fixing that appears a deep refactoring.

cleishm commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/EqualsTest.scala:
```diff
@@ -25,6 +25,57 @@ import org.scalatest.junit.JUnitRunner
+    // the parameter is used to disable auto-parametrization
+    val queryWithInt =
+      """PROFILE
+        |WITH {parameter} AS p
```
Might be nice to add a planner option like `PROFILE autoparam=false ...`, which is explicit, rather than faking it by having a dummy parameter. Something to consider.

Mats-SX replied:
Good idea, noted.

cleishm commented:
@craigtaverner Ok. Although I still suggesting making it equivalent to what equals does, which is assert that the comparison is against CTAny against CTAny. That means the equivalent for `IN` would be CTAny against CTCollection(CTAny). In this case, you'd make that line `rhs.expectType(CTCollection(CTAny).covariant) chain`, which I happen to have already tested and verified passes the new tests you wrote.

cleishm commented:
https://github.com/cleishm/neo4j/commit/85b91a8f3d3cc3a9e16b35af10f1684c74f932d8#diff-0a3a6439814d007f9ea8eb9518e61be2L103

Mats-SX commented:
@cleishm I implemented your fix in the PR and added a few tests that make sure that the semantic checking does not allow wrong types on the right-hand side, which our initial hack would've.

cleishm commented:
@Mats-SX: +1
